### PR TITLE
✨ zb,zm: Support Option<Header<'_>> parameter in property getters

### PR DESCRIPTION
**Pull request by rzvncj**
This allows for fine-grained connecting-UID checking, in order to be able to control who is able to access property values.

### CODE REVIEW

File: zbus_macros/src/iface.rs [hunk 1]
```diff
@@ -103,8 +103,8 @@ enum MethodType {
 
 #[derive(Debug, PartialEq, Copy, Clone)]
 enum PropertyType {
-    Inputs,
-    NoInputs,
+    Setter,
+    Getter,
```
Comment (zeenix) [inline zbus_macros/src/iface.rs:107]: I've been meaning to do this for a while now. Thanks for doing it. :)

File: zbus_macros/src/iface.rs [hunk 2]
```diff
@@ -1089,6 +1115,36 @@ fn introspect_method(name: &str, args: &TokenStream) -> TokenStream {
     )
 }
 
+fn is_special_arg(attrs: &[Attribute]) -> bool {
```
Comment (zeenix) [inline zbus_macros/src/iface.rs:1118]: nit: I've been trying to follow top-down code layout approach, which means use before definition. If you could just move it down after its first use, that would be great.

File: zbus/tests/e2e.rs [hunk 3]
```diff
@@ -257,7 +257,7 @@ impl MyIface {
 
     #[instrument]
     #[zbus(property)]
-    fn count(&self) -> u32 {
+    fn count(&self, #[zbus(header)] hea: Option<Header<'_>>) -> u32 {
```
Comment (zeenix) [inline zbus/tests/e2e.rs:260]: Kudos for adding a test case.

```suggestion
    fn count(&self, #[zbus(header)] header: Option<Header<'_>>) -> u32 {
```

Also let's do this in another (perhaps new) property and actually use the header. (you don't get a warning from the compiler about unused param? :thinking:)

File: zbus/src/object_server/interface/mod.rs [hunk 4]
```diff
@@ -45,10 +48,19 @@ pub trait Interface: Any + Send + Sync {
     }
 
     /// Get a property value. Returns `None` if the property doesn't exist.
```
Comment (zeenix) [inline zbus/src/object_server/interface/mod.rs:50]: Perhaps it'd be good to document here briefly when the `header` is `Some` and when it's `None`.

File: zbus_macros/src/iface.rs [hunk 5]
```diff
@@ -852,6 +865,9 @@ pub fn expand(args: Punctuated<Meta, Token![,]>, mut input: ItemImpl) -> syn::Re
             async fn get(
                 &self,
                 property_name: &str,
+                s: &#zbus::ObjectServer,
+                c: &#zbus::Connection,
+                h: &Option<#zbus::message::Header<'_>>,
```
Comment (zeenix) [inline zbus_macros/src/iface.rs:870]: I think it was a bad idea to choose such short non-descriptive names in the beginning but now that we're getting more args, it's even worse now. You're just following the existing code, so not at all your fault but if could you use the opportunity to change the code to use more meaningful names, that would be awesome. Bonus points for first updating the existing use in a separate commit that comes before this one. :wink: 

```suggestion
                object_server: &#zbus::ObjectServer,
                c: &#zbus::Connection,
                h: &Option<#zbus::message::Header<'_>>,
```

**Review by zeenix — COMMENTED** [review]
Really impressive work. I know it's quite difficult to get into the very complex and strange macro code and most contributors don't touch it.

Apart from the inline comments (mostly minor nitpicks), please add the new parameter to the `interface` docs and also to the book's `service` chapter.

**Review by zeenix — CHANGES_REQUESTED** [review]
oops, used the wrong option. :)

File: zbus/tests/e2e.rs (hunk 3 above)
Comment (rzvncj) [reply zbus/tests/e2e.rs:260]: No unused parameter warning, no. Maybe Rust is more permissive with `Option`s? Not sure.

**Comment by rzvncj** [comment]
Thanks for the prompt and very helpful review, and the kind words!

I've hopefully addressed your comments, other than updating the book and interface docs, and doing the "prequel" commit renaming the short parameter names - I'll be updating the PR with those changes as soon as possible too.

**Comment by rzvncj** [comment]
All done, I think, except for updating the book, because I'm not sure where this new information fits. Would you like me to update the `MyGreeter` example? Or a specific section? Or both?

**Comment by zeenix** [comment]
> All done, 

:+1: 

> I think, except for updating the book, because I'm not sure where this new information fits. Would you like me to update the `MyGreeter` example? Or a specific section? Or both?

I just had a look a the book again and seems we don't introduce all the special args there so never mind.

File: zbus_macros/src/iface.rs [hunk 6]
```diff
@@ -243,11 +243,11 @@ impl MethodInfo {
             let ret = quote!(r);
 
             quote!(match reply {
-                ::std::result::Result::Ok(r) => c.reply(&hdr, &#ret).await,
-                ::std::result::Result::Err(e) => c.reply_dbus_error(&hdr, e).await,
+                ::std::result::Result::Ok(r) => connection.reply(&hdr, &#ret).await,
+                ::std::result::Result::Err(e) => connection.reply_dbus_error(&hdr, e).await,
```
Comment (zeenix) [inline zbus_macros/src/iface.rs:260]: This commit does not touch `zbus` subcrate so `zb` prefix isn't valid here. :)

File: zbus/src/object_server/interface/mod.rs [hunk 7]
```diff
@@ -60,8 +77,9 @@ pub trait Interface: Any + Send + Sync {
         property_name: &'call str,
         value: &'call Value<'_>,
         emitter: &'call SignalEmitter<'_>,
+        header: &'call message::Header<'_>,
```
Comment (zeenix) [inline zbus/src/object_server/interface/mod.rs:80]: While this currently can always get a header passed to it, this will hopefully be changing at some point: https://github.com/dbus2/zbus/issues/898 . So best to make this optional as well. It will also be more consistent that way.

File: zbus_macros/src/lib.rs [hunk 8]
```diff
@@ -267,7 +267,7 @@ pub fn proxy(attr: TokenStream, item: TokenStream) -> TokenStream {
 ///   important. You can use `out_args` to specify their names.
 ///
 /// * `proxy` - Use this to specify the [`macro@proxy`]-specific method sub-attributes (e.g
-///   `object`). The common sub-attributes (e.g `name`) are automatically forworded to the
+///   `object`). The common sub-attributes (e.g `name`) are automatically forwarded to the
```
Comment (zeenix) [inline zbus_macros/src/lib.rs:270]: I appreciate your fixing the typo here but it's not relevant to this commit so if you could split it into a separate commit, that would be great (`git show --oneline && git reset HEAD^ && git add -p && git commit && git commit -a -C <commit-hash-from-git-show-command` should do it). :)

File: zbus_macros/src/lib.rs [hunk 9]
```diff
@@ -289,6 +289,13 @@ pub fn proxy(attr: TokenStream, item: TokenStream) -> TokenStream {
 /// using this since it will force all interested peers to fetch the new value and hence result in
 /// excess traffic on the bus.
 ///
+/// The getter property arguments support the following `zbus` attributes:
+///
+/// * `header` - This marks the property argument to receive an optional message header associated
+///   with the D-Bus property get operation being handled. It will be `None` when the getter is
+///   being called as part of the initial object setup (before it gets registered on the bus), or
+///   when we send out property changed notifications.
```
Comment (zeenix) [inline zbus_macros/src/lib.rs:297]: Good job but I think this doesn't need a special section. We should just add to the `header` arg docs below that in case of property methods, the header is optional and when its the case.

**Review by zeenix — CHANGES_REQUESTED** [review]
LGTM otherwise. Getting very close now. :)

**Comment by rzvncj** [comment]
> LGTM otherwise. Getting very close now. :)

Great, thanks! Submitted a new round of patches for review.

File: zbus_macros/src/lib.rs [hunk 10]
```diff
@@ -289,16 +289,19 @@ pub fn proxy(attr: TokenStream, item: TokenStream) -> TokenStream {
...
-/// The method arguments support the following `zbus` attributes:
+/// The method and property getter arguments support the following `zbus` attributes:
 ///
 /// * `object_server` - This marks the method argument to receive a reference to the
 ///   [`ObjectServer`] this method was called by.
 /// * `connection` - This marks the method argument to receive a reference to the [`Connection`] on
 ///   which the method call was received.
 /// * `header` - This marks the method argument to receive the message header associated with the
-///   D-Bus method call being handled.
+///   D-Bus method call being handled. For reading properties, this will be an `Option<Header<'_>>`,
```
Comment (zeenix) [inline zbus_macros/src/lib.rs:299]: > For reading properties

"For property getter methods"

**Review by zeenix — APPROVED** [review]
